openshift-ci/README.md

```diff
@@ -0,0 +1,193 @@
+
+### operator-registry-base
+
+The `operator-registry` is used as the base image for creating
+`operator-registry-base` image.  This is an intermediate image used to
+propagate value of `OPENSHIFT_BUILD_NAMESPACE` environment variable to
+`devconsole-operator-registry` image build.  This intermediate image
+is going to be used as the base image for
+`devconsole-operator-registry`.  The original `operator-registry`
+image has a a different value for `OPENSHIFT_BUILD_NAMESPACE`
+environment variable, which must be set when that image was built.
+
+The Dockerfile used is available here:
+`openshift-ci/Dockerfile.registry.intermediate` As you can see, the
+Dockerfile file has only one line with a `FROM` statement, which is
+going to replaced with `operator-registry` image during image build.
+
+### devconsole-operator-registry
+
+The `operator-registry-base` is used as the base image for creating
+`devconsole-operator-registry` image.
+
+The Dockerfile used is available here:
+`openshift-ci/Dockerfile.registry.build`
+
+The image produced can be pulled from here:
+`registry.svc.ci.openshift.org/${OPENSHIFT_BUILD_NAMESPACE}/stable:devconsole-operator-registry`
+
+## tests
+
+### lint
+
+The lint runs the GolangCI Lint, YAML Lint and Operator Courier.
+GolangCI is a Go program, whereas the other two are written in Python.
+So, Python 3 is a perquisite to run lint.
+
+The GolangCI Lint program runs multiple Go lint tools against the
+repository.  GolangCI Lint is runs lint concurrently and complete
+execution in few seconds. But there is one caveat, it requires lots of
+memory.  The memory limit has been increased to 6GB to accommodate the
+requirement.  As of now there is no configuration provided to run
+GolangCI Lint.
+
+The YAML Lint tools validate all the YAML configuration files.  It
+excludes the `vendor` directory while running.  There is a
+configuration file at top-level directory of the source code:
+`.yamllint`.
+
+The Operator Courier checks for the validity of Operator Lifecycle
+Manager (OLM) manifests.  That includes Cluster Service Version (CSV)
+files and CRDs.
+
+### test
+
+The `test` target runs the unit tests.  Some of the test make use mock
+objects. The unit tests doesn't require a dedicated OpenShift cluster
+unlike end-to-end tests.
+
+### e2e
+
+The `e2e` run an end-to-end test against an operator running inside
+the CI cluster pod but connected to a freshly created temporary
+Openshift 4 cluster.  It makes use the `--up-local` option provided by
+the Operator SDK testing tool.  It runs `test-e2e` Makefile target.
+
+### e2e-ci
+
+The `e2e-ci` target runs the end-to-end test against operator running
+inside the freshly created OpenShift 4 cluster.  All the resources are
+created through scripts invoked from `test-e2e-ci` Makefile target.
+This Makefile target is designed to run exclusively on CI environment.
+
+### e2e-olm-ci
+
+The e2e-ci runs the end-to-end test against operator running inside
```

Comment:
Should be "an operator"

Reply:
Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

Reply:
Pardon my ignorance, but are the e2e-ci tests a subset of e2e-olm-ci? What problems the former is expected to reveal that the latter cannot?

Reply:
The difference between `e2e-ci` and `e2e-olm-ci` is how the Kubernetes resources getting created. The resources include operator deployment, service account, role & role binding. The `e2e-ci` target uses a Makefile script (few `oc apply` commands) to create the resources whereas `e2e-olm-ci` uses Operator Lifecycle Manager (OLM) to create the resources.

Reply:
Thanks @baijum for helping me understand. Can you also answer the second part of my question?

Reply:
@rgarg1 Regarding revealing issues of the operator itself, there is no difference between these two end-to-end tests. The "e2e-olm-ci" verifies the Cluster Service Version file is correct, which is not happening in the previous "e2e-ci" test. However, the "e2e-olm-ci" setup is more complex, and there are more failure points. So having a simple setup to run end-to-end test would be helpful to iterate faster, even if there is some issue with OLM.